src/k2/common/FormattingUtils.h

```diff
@@ -0,0 +1,130 @@
+    SOFTWARE.
+*/
+#pragma once
+#include <fmt/format.h>
+
+#include <stdexcept>
+#include <type_traits>
+
+#include "AutoGenFormattingUtils.h"
+
+// helper function for converting enum class into an integral type
+// e.g. usage: auto integralColor = to_integral(MyEnum::Red);
+// or  std::array<MyEnum, to_integral(MyEnum::Red)>;
+template <typename T>
+inline auto to_integral(T e) { return static_cast<std::underlying_type_t<T>>(e); }
+
+// Intrusive formatting for classes
+// Generates formatting and from/to json conversion methods for a given CType, in a given namespace
+// e.g. usage:
+// namespace k2::dto {
+// struct GetSchemaRequest() {
+//    String schemaName;
+//    K2_DEF_FMT(GetSchemaRequest, schemaName)
+// };
+// }
+#define K2_DEF_FMT(CType, ...)                                                                    \
+    template <typename OStream_T>                                                                 \
+    friend OStream_T& operator<<(OStream_T& os, const CType& o) {                                 \
+        if constexpr (std::is_same<OStream_T, std::ostream>::value) {                             \
+            fmt::print(os,                                                                        \
+                           FMT_STRING("[" _K2_MKLIST(__VA_ARGS__) "]") _K2_MKVARS(__VA_ARGS__));  \
+        } else {                                                                                  \
+            fmt::format_to(os.out(),                                                              \
+                           FMT_COMPILE("[" _K2_MKLIST(__VA_ARGS__) "]") _K2_MKVARS(__VA_ARGS__)); \
+        }                                                                                         \
+        return os;                                                                                \
+    }                                                                                             \
+    friend void to_json(nlohmann::json& j, const CType& o) {                                      \
+        (void)o;                                                                                  \
+        j = nlohmann::json{_K2_TO_JSON(__VA_ARGS__)};                                             \
+    }                                                                                             \
+                                                                                                  \
+    friend void from_json(const nlohmann::json& j, CType& o) {                                    \
+        _K2_FROM_JSON(__VA_ARGS__)                                                                \
+    }
+
+// Generating enums with formatting. the _IC version is to be used when inside a class.
+// e.g.
+// namespace k2::dto {
+// K2_DEF_ENUM(TxnStatus, Created, Aborted, Committed) )
+// }
+#define K2_DEF_ENUM_IC(Name, ...)                                                   \
```

Comment:
Probably should use something more specific than "Name" to prevent aliasing

Reply:
done